# rRNA_counter
Count and compare 16S rRNA genes within a genome

Takes an input genus and returns a large set of analysed files, the main being phylogenetic trees of full length 16S rRNA genes and common amplicons thereof.


## Installation

#Clone from github

git clone https://github.com/mikaells/rRNA_counter/

#Create conda environment form yml file

conda env create -f rRNA_counter/rRNA_counter.yml

#Activate conda environment 

conda activate rRNA

#Set permissions of files to be executable 

chmod 755 rRNA_counter/*

#silence the parallel citation message

parallel --citation   


Note: in_silico_PCR.pl is borrowed from https://github.com/egonozer/in_silico_pcr


## Usage
#Run analysis by as follows 

path-to-rRNA_counter/rRNA_counter.sh -g $genus

#An example run with Ruegeria as target genus:

~/rRNA_counter/rRNA_counter.sh -g Ruegeria

#Can run with species as well, make sure to use quotes

~/rRNA_counter/rRNA_counter.sh -g "Mycoplasma bovis"

## Output

The program generates a new directory named after the genus in question. Within that is a summary file and three sub directories:

$genus-summary.csv # contains summary statistics for each genome. Columns 5-8 are in nucleotide mismatches between the genomes' 16S genes, TotalDiv is the cummulative shannon index
 
    #GCF     Genus   Species #16S    Mean    SD      Min     Max   TotalDiv 

refseq/bacteria/ #contains a folder for each genome, named after refseq-identifier

    GCF*.fna # the original genome file, headers are renamed for downstream compatibility
  
    GCF*.fna.fai # indexed fna file
  
    GCF*.rRNA # rRNA file generated by barrnap
  
    GCF*.16S # only the 16S file from .rRNA file
  
    GCF*.16sAln # alligned .16S file (by muscle)
  
    GCF*.16sTree # tree from .16sAln file (by fasttree)
  
    16S_div.pdf # pdf for 16S tree and diversity across genes
  
    indiv_16S_dir/ # folder for individual 16S files, for pyani compatibility
  
    x.fna #16s files
  
    ani/ # files generated from pyani


full/ # files for full length 16S analysis
  
    $genus.16S # all 16S genes
  
    $genus.aln # alligned from 16S genes (alligned by mafft)
  
    $genus.tree # tree from allignment
  
    $genus-CLC.csv # summary file for import into CLC
    

amplicons/ #files for amplicons
  
    $genus-$region.16S # all 16S genes
  
    $genus-$region.aln # alligned from 16S genes (alligned by mafft)
  
    $genus-$region.tree # tree from allignment
  
    $genus-$region-CLC.csv # summary file for import into CLC
    
  
The main files of interest are the trees from full/ and amplicons/ as well as the $genus-summary.csv file.

The tree files can be used in any treeviewer or R. The CLC-csv file is formatted for CLC, but is just a table and can be used for other treeviewers supporting annotation.
